symengine/pow.cpp

```diff
@@ -112,61 +112,60 @@ RCP<const Basic> pow(const RCP<const Basic> &a, const RCP<const Basic> &b)
-                        == 2)) {
+        } else if (is_a<Rational>(*b) and eq(*b, *rational(1, 2))) {
             return I;
         }
     }
 
-    if (is_a_Number(*a) and is_a_Number(*b)) {
-        if (is_a<Integer>(*b)) {
-            if (is_a<Rational>(*a)) {
-                RCP<const Rational> exp_new
-                    = rcp_static_cast<const Rational>(a);
-                return exp_new->powrat(*rcp_static_cast<const Integer>(b));
-            } else if (is_a<Integer>(*a)) {
-                RCP<const Integer> exp_new = rcp_static_cast<const Integer>(a);
-                return exp_new->powint(*rcp_static_cast<const Integer>(b));
-            } else if (is_a<Complex>(*a)) {
-                RCP<const Complex> exp_new = rcp_static_cast<const Complex>(a);
-                RCP<const Integer> pow_new = rcp_static_cast<const Integer>(b);
-                RCP<const Number> res = exp_new->pow(*pow_new);
-                return res;
-            } else {
+    if (is_a_Number(*b)) {
+        if (is_a_Number(*a)) {
+            if (is_a<Integer>(*b)) {
                 return down_cast<const Number &>(*a)
                     .pow(*rcp_static_cast<const Number>(b));
-            }
-        } else if (is_a<Rational>(*b)) {
-            if (is_a<Rational>(*a)) {
-                return down_cast<const Rational &>(*a)
-                    .powrat(down_cast<const Rational &>(*b));
-            } else if (is_a<Integer>(*a)) {
-                return down_cast<const Rational &>(*b)
-                    .rpowrat(down_cast<const Integer &>(*a));
-            } else if (is_a<Complex>(*a)) {
+            } else if (is_a<Rational>(*b)) {
+                if (is_a<Rational>(*a)) {
+                    return down_cast<const Rational &>(*a)
+                        .powrat(down_cast<const Rational &>(*b));
+                } else if (is_a<Integer>(*a)) {
+                    return down_cast<const Rational &>(*b)
+                        .rpowrat(down_cast<const Integer &>(*a));
+                } else if (is_a<Complex>(*a)) {
+                    return make_rcp<const Pow>(a, b);
+                } else {
+                    return down_cast<const Number &>(*a)
+                        .pow(*rcp_static_cast<const Number>(b));
+                }
+            } else if (is_a<Complex>(*b)
+                       and down_cast<const Number &>(*a).is_exact()) {
                 return make_rcp<const Pow>(a, b);
             } else {
                 return down_cast<const Number &>(*a)
                     .pow(*rcp_static_cast<const Number>(b));
             }
-        } else if (is_a<Complex>(*b)) {
-            return make_rcp<const Pow>(a, b);
-        } else {
-            return down_cast<const Number &>(*a)
-                .pow(*rcp_static_cast<const Number>(b));
+        } else if (is_a<Constant>(*a)) {
+            RCP<const Number> p = rcp_static_cast<const Number>(b);
+            if (not p->is_exact()) {
```

Comment:
@jppelteret The function will return `make_rcp<const Pow>(a, b);`. Basically, it will be printed as it is, no numeric evaluation would be done. `not p->is_exact()` statement holds true for `RealDouble` and similar classes, as a double is never exact.  @isuruf Please correct me if I am mistaken.